GH-40339: [Java] StringView Initial Implementation

vibhatha opened the pull request:
### Rationale for this change 

StringView implementation in Java. This PR only includes the core implementation of StringView

### What changes are included in this PR?

- [X] Adding ViewVarBinaryVector
- [X] Adding ViewVarCharVector
- [X] Adding corresponding test cases in the given scope
- [X] Including required implementation extensions with not supported warnings
- [X] Interface for Holders

### Non Goals of this PR

- [ ] https://github.com/apache/arrow/issues/40937
- [ ] https://github.com/apache/arrow/issues/40936
- [ ] https://github.com/apache/arrow/issues/40932
- [ ] https://github.com/apache/arrow/issues/40943
- [ ] https://github.com/apache/arrow/issues/40944
- [ ] https://github.com/apache/arrow/issues/40942
- [ ] https://github.com/apache/arrow/issues/40945
- [ ] https://github.com/apache/arrow/issues/40941
- [ ] https://github.com/apache/arrow/issues/40946

### Are these changes tested?

Yes. Existing test cases on `VarCharVector` and `VarBinaryVector` are verified with view implementations and additional test cases have also been added to check view functionality. And explitly tests have been added to evaluate the view functionality with `ViewVarCharVector`

### Are there any user-facing changes?

Yes, this introduces a new API and some public methods have been included in an interface so that it can be extended to write custom functionality like done for views. 

* GitHub Issue: #40339

lidavidm reviewed (commented):
Some quick comments. I'm not done looking at this yet.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/compare/TypeEqualsVisitor.java:
```diff
@@ -85,6 +86,11 @@ public Boolean visit(BaseLargeVariableWidthVector left, Void value) {
 
+  @Override
+  public Boolean visit(BaseVariableWidthViewVector left, Void value) {
+    throw new UnsupportedOperationException("View vectors are not supported.");
```
It is not the difficulty, I actually thought about the size of this PR. It is already over 3000 LoC. I was merely trying move possible things to separate PRs.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/TypeLayout.java:
```diff
@@ -347,11 +360,23 @@ public Integer visit(Binary type) {
       }
 
+      @Override
+      public Integer visit(ArrowType.BinaryView type) {
```
Yes I agree, I have created an issue for this. How should we proceed?

lidavidm replied:
We need to look at how/where this is used.

vibhatha replied:
At the moment, I kept this unsupported and I already have created an issue to follow up. I am also not sure about this. But if we think about the layout, and keep the growing external buffer as K variable, the base minimum number of buffers can be defined as 2, which is validity and view buffer. But I am not sure if that actually make sense as layout should have that external buffer represented in some way. Or can we include a callback-like thing to get the precise value? But that is again a bit odd as layout supposed to be like an enum-like class. 

Not exactly sure what to do here. 

// TODO: https://github.com/apache/arrow/issues/40934

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1585 @@
+  private long lastValueAllocationSizeInBytes;
+
+  public static final int OFFSET_WIDTH = 4; /* 4 byte unsigned int to track offsets */
+  /* protected members */
```
nit: redundant comment

vibhatha replied:
removed

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1585 @@
+  private int lastValueCapacity;
+  private long lastValueAllocationSizeInBytes;
+
+  public static final int OFFSET_WIDTH = 4; /* 4 byte unsigned int to track offsets */
```
er, we don't have an offset buffer?

vibhatha replied:
You're correct. I will fix this sooner.

vibhatha replied:
removed.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1585 @@
+
+  public static final int OFFSET_WIDTH = 4; /* 4 byte unsigned int to track offsets */
+  /* protected members */
+  protected static final int INLINE_SIZE = 12; /* 12 byte unsigned int to track inline views*/
```
Can you use proper doc comment syntax...?

vibhatha replied:
Updated the comment section.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1585 @@
+  protected static final int VIEW_BUFFER_SIZE = 16; /* 16 byte default size for each view*/
+  protected static final int LENGTH_WIDTH = 4; /* the first 4 bytes of view are allocated for length*/
+  protected static final int PREFIX_WIDTH = 4; /* the second 4 bytes of view are allocated for prefix width*/
+  protected static final int BUF_INDEX_WIDTH = 4; /* third 4 bytes of view are allocated for buffer index*/
```
Honestly, these constants don't line up with the spec. Where is the constant for length?

vibhatha replied:
I will revisit this. And properly define with better naming.

vibhatha replied:
@lidavidm is it still unclear?

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1585 @@
+  protected static final int BUF_INDEX_WIDTH = 4; /* third 4 bytes of view are allocated for buffer index*/
+  protected static final byte[] emptyByteArray = new byte[]{};
+  protected ArrowBuf validityBuffer;
+  protected ArrowBuf valueBuffer;
```
We have more than one values buffer though?

vibhatha replied:
I think I should rename this. But before that let me explain the way I implemented the views. Ignoring the incorrect offset, the rest is as follows. 

The valueBuffer is basically the views.
It contains short string inline and long strings with the length, prefix, buf index and offset. The list of ArrowBuf a.k.a dataBuffers are the buffers which keeps the full strings for long strings. This buffer has a default size. Once it's filled to a certain limit where next long string cannot be stored a new ArrowBuf is added to this list. So the first element of the dataBuffer would be index 0 and beyond that 1, and then 2 after adding the third buffer. 

Is this incorrect? Is there a better alternative?

lidavidm replied:
As long as it's clearly documented.

lidavidm replied:
But as noted the spec calls it the "views" buffer.

vibhatha replied:
Of course my mistake for not properly naming and commenting. Is the way I used follow the spec? Sorry about the confusion.

vibhatha replied:
Renamed.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1585 @@
+  protected int valueCount;
+  protected int lastSet;
+  protected final Field field;
+  protected List<ArrowBuf> dataBuffers;
```
...why do we have both this and "valueBuffer"?

vibhatha replied:
Explained here: https://github.com/apache/arrow/pull/40340#discussion_r1551080873

Please correct me if I am wrong.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1585 @@
+  protected static final byte[] emptyByteArray = new byte[]{};
+  protected ArrowBuf validityBuffer;
+  protected ArrowBuf valueBuffer;
+  protected ArrowBuf offsetBuffer;
```
The spec calls the second buffer the "views" buffer.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1585 @@
+    super(allocator);
+    this.field = field;
+    lastValueAllocationSizeInBytes = INITIAL_BYTE_COUNT;
+    // -1 because we require one extra slot for the offset array.
```
This isn't true...We don't have offsets...

vibhatha replied:
Related question: 

Since offset is not a part of this. The number of elements the ViewVarcharVector can hold vs number of elements VarCharVector can hold should be different right?

lidavidm replied:
I mean yes? One is view_buffer / 16 and the other is offsets_buffer / offset_width - 1

lidavidm replied:
But I'm trying to say you should _not_ be referencing the VarCharVector as the base of the implementation here

vibhatha replied:
Yes indeed.

lidavidm reviewed (changes requested):
@vibhatha can you go over the base implementation again with the spec? It seems to be following the offsets buffer implementation still but none of that applies to view vectors.

vibhatha commented:
@lidavidm the offset usage is incorrect here, it's not needed. I need to fix that. Thanks a lot for catching that.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1570 @@
+   * @param density average number of bytes per variable width element
+   */
+  @Override
+  public void setInitialCapacity(int valueCount, double density) {
```
@lidavidm 
Thinking about this again, I feel like this method is not practical for views.

lidavidm replied:
It would set the capacity of the view buffer based on valueCount and possibly pre-allocate the data buffer based on valueCount * density.

vibhatha replied:
I am not sure will it be that useful though. At the moment, we allocate it on demand. In case of we only have short strings, this would be a kind of a memory waste? Or is there an alternative objective? 
I am actually trying to understand the purpose of this for views :thinking:

vibhatha replied:
Updated the functionality to keep an initialDataBufferSize, later when the allocation happens, we choose this value if provided for the allocation.

vibhatha commented:
> @vibhatha can you go over the base implementation again with the spec? It seems to be following the offsets buffer implementation still but none of that applies to view vectors.

@lidavidm I updated the code, I am still verifying the changes. I will ping you once it is ready for another round of reviews. 
Thanks for the support.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/AbstractVariableWidthVector.java:
```diff
@@ -0,0 +1,146 @@
+/**
+* A base class for common functionalities in variable width vectors.
+*/
+public interface AbstractVariableWidthVector extends VariableWidthVector, FieldVector, VectorDefinitionSetter {
```
@lidavidm is this name okay or should we go for something like `VariableWithVectorOperations` or any better suggestion?

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+  }
+
+  protected final void setBytes(int index, byte[] value, int start, int length) {
+    createViewBuffer(allocator, index, value, start, length, viewBuffer, dataBuffers);
```
keep it as it is or, move that whole functionality within? 
Only used within this class and in one of the test cases.

lidavidm replied:
Why are we passing arguments to it like allocator or dataBuffers? They're part of the class.

vibhatha replied:
Wrote that function considering passing an allocator provided by the user. But we can remove it, and I understand your point.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+   * @param index The index of the element in the vector.
+   * @return The total length up to the element at the given index.
+   */
+  public final int getTotalLengthUptoIndex(int index) {
```
This is a useful function, but I have doubts about the naming.

lidavidm replied:
```suggestion
  public final int getTotalValueLengthUpToIndex(int index) {
```

vibhatha commented on java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java:
```diff
@@ -2019,89 +2118,102 @@ public void testCopyFromWithNulls1() {
+    assertEquals(0, vector.getValueLength(19));
+
+    /* Check offsets */
+    if (vector instanceof BaseVariableWidthVector) {
```
I did this to minimize the duplicated content. Though I have doubts isn't this a bit hard to read?  Instead of doing this would it be okay if I use a separate method for each vector representation, or maybe add an external function for offset verification in case of `BaseVariableWidthVector`?

lidavidm replied:
I don't see what the problem is necessarily?

vibhatha replied:
I think it would be better to have something like `commonTestXYZ` and do the assertions in a vector specific manner. 
You have pointed it out here: https://github.com/apache/arrow/pull/40340#discussion_r1557121026

vibhatha replied:
Moved the vector specific tests as explained above.

vibhatha commented on java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java:
```diff
@@ -2167,140 +2279,179 @@ public void testVectorLoadUnload() {
+    assertEquals(0, vector.getValueLength(14));
+
+    /* Check offsets */
+    if (vector instanceof BaseVariableWidthVector) {
```
same as above, is this okay or keep separate methods for each vector type with certain degree of duplication?

vibhatha replied:
This has been resolved.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+ */
+public abstract class BaseVariableWidthViewVector extends BaseValueVector implements AbstractVariableWidthVector {
+  // A single element of a view comprises 16 bytes
+  protected static final int VIEW_BUFFER_SIZE = 16;
```
is this okay or a better name for this?

lidavidm replied:
probably technically ELEMENT_SIZE or something so that it's clear?

vibhatha replied:
Sure, I will refactor.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+public abstract class BaseVariableWidthViewVector extends BaseValueVector implements AbstractVariableWidthVector {
+  // A single element of a view comprises 16 bytes
+  protected static final int VIEW_BUFFER_SIZE = 16;
+  public static final int INITIAL_VIEW_VALUE_ALLOCATION = 4096;
```
since we don't split offset and validity, I thought it is better to you this instead of `3970`, any thoughts?

lidavidm replied:
rounding up sounds fine

lidavidm replied:
not sure where 3970 comes from anyways?

vibhatha replied:
https://github.com/apache/arrow/blob/main/java/vector/src/main/java/org/apache/arrow/vector/BaseValueVector.java#L41-L47
I was referring to this.

vibhatha replied:
I understand.

vibhatha commented on java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java:
```diff
@@ -2167,140 +2279,179 @@ public void testVectorLoadUnload() {
+  }
+
+  @Test /* VarCharVector */
+  public void testGetBufferAddress1() {
```
here I did split for readability.

vibhatha replied:
One for `ViewVarCharVector` and the other for `VarCharVector`

lidavidm replied:
IMO it is less readable that I have to read two very similar blocks of code and jump around. A single use method is not necessarily clearer than just having a long method unless you have some opportunity to document things or something.


If you are worried about the vector-specific code I would do

```java
    try (final VarCharVector vector = new VarCharVector("myvector", allocator)) {
      commonTest(vector);
      assertEquals(/*...type specific tests go here...*/);
    }

    try (final ViewVarCharVector vector = new ViewVarCharVector("myviewvector", allocator)) {
      commonTest(vector);
    }
```

vibhatha replied:
Addressed as advised.

vibhatha commented on java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java:
```diff
@@ -2393,6 +2561,48 @@ public void testSetInitialCapacity() {
   }
 
+  @Test /* ViewVarCharVector */
+  public void testSetInitialCapacityInViews() {
```
@lidavidm I have a few doubts about this test case, appreciate a closer look. 
The allocation behavior for views and non-views are a bit different since we have the 16bytes offset.

vibhatha replied:
I updated the test case to have a completely separate test criterion.

vibhatha commented:
@lidavidm I updated the code and added some comments. Appreciate another round of reviews.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+  }
+
+  /**
+   * Get the size (number of bytes) of underlying data buffer.
```
Let's be consistent and call this the view buffer, given the data buffer is separate?

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+        // the first data buffer needs to be added
+        ArrowBuf newDataBuf = allocator.buffer(referenceBufferSize);
+        // set length
+        valueBuffer.setInt(writePosition, length);
```
factor out the code dealing with the view buffer? you should first allocate a buffer if needed or else just take the last buffer, then append to the buffers. don't copy-paste the code.

lidavidm replied:
also, please fix the naming...let's use "view buffer" consistently here. even if some of the API methods still call it the "value buffer", we can at least use the right name internally.

vibhatha replied:
Ah yes. Sorry about that.

vibhatha replied:
I updated the code, please take a look.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+    return totalLength;
+  }
+
+  private static long roundUpToMultipleOf16(long num) {
```
move to DataSizeRoundingUtil

vibhatha commented on java/vector/src/main/codegen/templates/ValueHolders.java:
```diff
@@ -40,11 +42,18 @@ public final class ${className} implements ValueHolder{
     <#else>
+
+    <#if minor.class == "ViewVarChar" || minor.class == "ViewVarBinary">
+    public HolderCallback callBack;
```
I added this interface for defining the holders required for set and get methods. Basically, for non views we can just add offsets, valueBuffer and retrieve the value later or set it. For views the retrieval and set logic are a bit complicated than that. So I use this callback to basically encapsulate that detail from the user.

lidavidm replied:
The holder encapsulates that detail itself.

lidavidm replied:
Also (1) this isn't a callback, and (2) it's entirely undocumented anyways

lidavidm replied:
I don't see why the fields can't directly be in the holder.

vibhatha replied:
Not the fields the way to get the required value? In other cases knowing offset and length is sufficient. But here we have to retrieve buf index, offset and then get the value. Do we expect user to do it themselves? Am I understanding the holder wrong way here?

lidavidm replied:
The holder is a fairly thin wrapper.

In any case, I would actually expect this to just be a single buffer containing the data. It'll be either sliced from the views buffer or a data buffer.

lidavidm replied:
Look at VarCharHolder: it's just a buffer and a start/end. If anything, that holder already works for this vector.

vibhatha replied:
Yes pretty much slicing but user needs to know the spec. So let's assume they do and keep this as lightweight as possible?

lidavidm replied:
The lightest weight thing would just be a buffer and a start/end.

vibhatha replied:
Sure. I will update.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+  /**
+   * Get the inner vectors.
+   *
+   * @deprecated This API will be removed as the current implementations no longer support inner vectors.
```
Sure I can do that.

vibhatha replied:
@lidavidm we have two general filings 
1. https://github.com/apache/arrow/issues/15167
2. https://github.com/apache/arrow/issues/33022

I think [1] suits better. I can list this one there, but it would be better if we do the investigation and list all deprecated warnings and remove them in a graceful manner.

vibhatha replied:
Issue for this: https://github.com/apache/arrow/issues/41069

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+  /**
+   * Reallocate the data buffer for reference buffer.
+   */
+  public void reallocViewReferenceBuffer() {
```
I think this would be my bad naming. What I meant was the external buffer which keeps the long strings. I used the reference, as the viewBuffer references this via offset and index.

lidavidm replied:
The spec calls this the data buffer. Let's stick to the names there.

vibhatha replied:
Sure @lidavidm 
I saw the term data buffer in a util where we use a single buffer, split it and make validity and offset buffers. Just curious, would that need a renaming? As it was a bit confusing. Not related to this in anyway, but just for understanding.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+  /**
+   * zero out the vector and the data in associated buffers.
+   */
+  public void zeroVector() {
```
Yes of course.

vibhatha replied:
Thanks for catching this.

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+    // for details, please see TestValueVector#testUnloadVariableWidthVector.
+    fillHoles(valueCount);
+
+    List<ArrowBuf> result = new ArrayList<>(2);
```
```suggestion
    List<ArrowBuf> result = new ArrayList<>(2 + dataBuffers.size());
```
right?

vibhatha commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1464 @@
+    // for each set operation, we have to allocate 16 bytes
+    // here we are adjusting the desired allocation-based allocation size
+    // to align with the 16bytes requirement.
+    newAllocationSize += VIEW_BUFFER_SIZE;
```
I need to check this one, @lidavidm

vibhatha replied:
Thanks David for catching this, I updated a related path in this line of function calls for memory allocation. 
Could you please take a look at the relloaction logic?

vibhatha commented:
@lidavidm ready for another round of reviews, also posted a few questions too.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1434 @@
+   * Validate the scalar values held by this vector.
+   */
+  public void validateScalars() {
+    // No validation by default.
```
Is there an issue filed for this? We should be able to validate the length and so on.

vibhatha replied:
Not specifically, this was the same case for the `VarCharVector` representation.

vibhatha replied:
I left it as it is for the `VarCharVector`, and this was a [recently added feature](https://github.com/vibhatha/arrow/commit/a0041027902ac71f00572d5b31022ee98502ed6a).

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1434 @@
+    if (value.length <= INLINE_SIZE) {
+      // allocate inline buffer
+      // set length
+      viewBuffer.setInt(writePosition, length);
```
this can be done outside the if/else?

vibhatha replied:
We can leave setting the length as a common part, but I would rather like to leave it as it is for readability in a sense that, for a short string this is how we allocate vs allocation for a long string. But if it needs moving, we can do, not a problem :slightly_smiling_face:

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1434 @@
+    return viewBuffer.getInt(((long) index * ELEMENT_SIZE));
+  }
+
+  protected ArrowBuf allocateOrGetLastBuffer(BufferAllocator allocator, int length, List<ArrowBuf> dataBuffers) {
```
```suggestion
  protected ArrowBuf allocateOrGetDataBuffer(BufferAllocator allocator, int length, List<ArrowBuf> dataBuffers) {
```

vibhatha replied:
Yes indeed, we don't need to pass any parameters here since they are all part of the class, thought of an additional use case to make it accessible to the package, but I can see your point, not relevant or required in the current context. 

I will remove the parameters.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1434 @@
+  }
+
+  protected ArrowBuf allocateOrGetLastBuffer(BufferAllocator allocator, int length, List<ArrowBuf> dataBuffers) {
+    long dataBufferSize;
+    if (initialDataBufferSize > 0) {
+      dataBufferSize = initialDataBufferSize;
+    } else {
+      dataBufferSize = lastValueAllocationSizeInBytes;
+    }
+
+    if (dataBuffers.isEmpty() || dataBuffers.get(dataBuffers.size() - 1).capacity() -
+            dataBuffers.get(dataBuffers.size() - 1).writerIndex() < length) {
+      ArrowBuf newBuf = allocator.buffer(dataBufferSize);
+      dataBuffers.add(newBuf);
+    }
```
Not so sure about this logic. Don't you want to reallocate existing buffers where possible? And, what if length is greater than dataBufferSize?

vibhatha replied:
> And, what if length is greater than dataBufferSize?

Correct we need to check the possible size with the length and allocate the adequate. But since, the dataBuffer can get more allocations, it would be better to allocate whatever the highest (`length` or `initialDataBufferSize` or `lastValueAllocationSizeInBytes`)

vibhatha replied:
>  Don't you want to reallocate existing buffers where possible? 

About position, I am merely checking how much space has been consumed in the last appeneded databuffer in the dataBuffers. Could you please explain a bit if I have understood this wrong?

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1434 @@
+   * @param index The index of the element in the vector.
+   * @return The length of the element at the given index.
+   */
+  protected final int getLength(int index) {
```
```suggestion
  protected final int getValueLength(int index) {
```

vibhatha replied:
We can remove this method. I have introduced a duplicated functionality. Sorry about that.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1434 @@
+   *----------------------------------------------------------------*/
+
+
+  protected final void fillHoles(int index) {
```
Why do we need to do this? I think this applies only to VarCharVector, and not views.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1434 @@
+   * @param viewBuffer The ArrowBuf instance that will hold the data of the new value.
+   * @param dataBuffers The list of ArrowBuf instances that hold the data of all long binary values in the vector.
+   */
+  protected void createViewBuffer(
```
```suggestion
  protected void setValue(
```

or something, but this is not about creating a "view buffer"

vibhatha replied:
@lidavidm btw, the objective of making this separate function has become obsolete after the refactor. 
I think we could just put the content to into `setBytes` and use it that way.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/ViewVarBinaryVector.java:
```diff
@@ -0,0 +1,230 @@
+   */
+  @Override
+  public void read(int index, ReusableBuffer<?> buffer) {
+    byte[] data = getData(index);
+    buffer.set(data, 0, data.length);
```
These implementations unnecessarily do an extra copy. We should be able to set directly from a backing buffer.

vibhatha replied:
I added a new method `getData(int index, ReusableBuffer<?> buffer)`, could you please check?

vibhatha commented on java/vector/src/main/codegen/templates/ValueHolders.java:
```diff
@@ -40,11 +39,17 @@ public final class ${className} implements ValueHolder{
+    <#if minor.class == "ViewVarChar" || minor.class == "ViewVarBinary">
+    public List<ArrowBuf> dataBuffers;
+    public ArrowBuf outputBuffer;
+    </#if>
```
I think my interpretation of a holder wasn't accurate. I have removed both of these attributes.

vibhatha commented on java/vector/src/test/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java:
```diff
@@ -586,6 +588,31 @@ public static void setVector(VarCharVector vector, byte[]... values) {
     vector.setValueCount(length);
   }
 
+  public static void setVector(VariableWidthFieldVector vector, byte[]... values) {
+    final int length = values.length;
+    vector.allocateNewSafe();
+    for (int i = 0; i < length; i++) {
+      if (values[i] != null) {
+        vector.set(i, values[i]);
+      }
+    }
+    vector.setValueCount(length);
+  }
+
+  /**
+   * Populate values for ViewVarCharVector.
+   */
+  public static void setVector(ViewVarCharVector vector, byte[]... values) {
+    final int length = values.length;
+    vector.allocateNewSafe();
+    for (int i = 0; i < length; i++) {
+      if (values[i] != null) {
+        vector.set(i, values[i]);
+      }
+    }
+    vector.setValueCount(length);
+  }
```
You're correct, this is wrong. Something I missed down the line when the common interface was introduced.

vibhatha commented:
@lidavidm I addressed the reviews and added a few comments too. Appreciate your feedback once again :slightly_smiling_face:

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1402 @@
+  }
+
+  /**
+   * Reset the vector to initial state. Same as {@link #zeroVector()}.
```
it looks like it is slightly different than zeroVector?

vibhatha replied:
I used the comment from `BaseVariableWidthVector`, didn't put much thought to it at the time. 
I agree with you. Let's remove that part.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1402 @@
+  }
+
+  @Override
+  public int sizeOfValueBuffer() {
```
This operation is ill-defined, I think. It's not clear to me what it's supposed to actually mean for this type.

vibhatha replied:
Thinking about this again... 
The interface definition for this function is 

```java
/**
   * Provide the number of bytes contained in the valueBuffer.
   * @return the number of bytes in valueBuffer.
   */
   int sizeOfValueBuffer();
```
We don't have a `valueBuffer` for views. I think what we could do is the following. 

First step, Check usage and avoid. This is only used in test cases and that is used with the abstracted test cases using the common interface to minimize test code duplication for views and non-views. The test case is trivial it just needs to know the size of the data stored. `sizeOfValueBuffer` usage can be avoided and we can just use the `getValueLength` and `getValueCount` functions to get the same functionality. 

In terms of functionality support, we should probably override this and put `UnsupportedOperationException`. And we can introduce a separate function to get the `number of bytes contained in the viewBuffer`, which is `16* valueCount`.

WDYT?

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1402 @@
+    assert index >= 0;
+    final int dataLength = end - start;
+    BitVectorHelper.setValidityBit(validityBuffer, index, isSet);
+    byte[] data = new byte[dataLength];
+    buffer.getBytes(start, data, 0, dataLength);
+    setBytes(index, data, start, dataLength);
```
We should be able to avoid this copy?

vibhatha replied:
We can introduce an overload like `protected final void setBytes(int index, ArrowBuf valueBuf, int start, int length)` and avoid that additional copy.

vibhatha replied:
There is a degree of code duplicates in the updated version. Although I tried to abstract it away as follows, but I got an exception.

Failed abstraction attempt 

```java
protected final void setBytes(int index, byte[] value, int start, int length) {
    ArrowBuf currentBuf = null;
    if (length > INLINE_SIZE) {
        currentBuf = allocateOrGetLastDataBuffer(length);
        currentBuf.setBytes(currentBuf.writerIndex(), value, start, length);
        currentBuf.writerIndex(currentBuf.writerIndex() + length);
    }
    setViewBuffer(index, value, start, length, currentBuf);
}

protected final void setBytes(int index, ArrowBuf valueBuf, int start, int length) {
    ArrowBuf currentBuf = null;
    if (length > INLINE_SIZE) {
        currentBuf = allocateOrGetLastDataBuffer(length);
        currentBuf.setBytes(currentBuf.writerIndex(), valueBuf, start, length);
        currentBuf.writerIndex(currentBuf.writerIndex() + length);
    }
    setViewBuffer(index, valueBuf.nioBuffer().array(), start, length, currentBuf);
}

private void setViewBuffer(int index, byte[] value, int start, int length, ArrowBuf currentBuf) {
    int writePosition = index * ELEMENT_SIZE;
    viewBuffer.setInt(writePosition, length);
    writePosition += LENGTH_WIDTH;
    if (length <= INLINE_SIZE) {
        viewBuffer.setBytes(writePosition, value, start, length);
    } else {
        viewBuffer.setBytes(writePosition, value, start, PREFIX_WIDTH);
        writePosition += PREFIX_WIDTH;
        viewBuffer.setInt(writePosition, dataBuffers.size() - 1);
        writePosition += BUF_INDEX_WIDTH;
        viewBuffer.setInt(writePosition, (int) currentBuf.writerIndex());
    }
}
```

```bash
java.lang.UnsupportedOperationException
	at java.base/java.nio.ByteBuffer.array(ByteBuffer.java:1505)
at org.apache.arrow.vector/org.apache.arrow.vector.BaseVariableWidthViewVector.setBytes(BaseVariableWidthViewVector.java:1245)
	at org.apache.arrow.vector/    org.apache.arrow.vector.BaseVariableWidthViewVector.setSafe(BaseVariableWidthViewVector.java:1115)
	at org.apache.arrow.vector/org.apache.arrow.vector.TestValueVector.testSetSafeWithArrowBufNoExcessAllocs(TestValueVector.java:1978)

```

failed at calling -> `valueBuf.nioBuffer().array()`

I wasn't sure, thus I settled for a bit code duplicated approach where `setBytes` has 2 versions which supports `byte[]` input and `ArrowBuf` as an input. Can I do better here?

lidavidm replied:
I don't think it matters that much to duplicate a ~dozen lines of code

lidavidm commented on java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java:
I think this is kind of a mess, and almost think we should just write all new tests for the view vectors, also because then we can make sure to test the different behavior of view vectors properly.

vibhatha replied:
I agree @lidavidm. So my idea was to make sure the following

1. Can `ViewVarCharVector` do what is being done with `VarCharVector`? I tried to cover this basis using the `TestValueVector` test cases. 
2. I added another test suite called `TestVarCharViewVector` to test the view-oriented cases? Maybe we can do more improvements there?

lidavidm replied:
Let's just split the tests entirely.

vibhatha replied:
Sure, I can do that. Also I added the common interface mainly for make it easier to write test cases. 
Should we also remove that interface -> `VariableWidthFieldVector`?

lidavidm replied:
I think the interface is fine.

lidavidm commented on java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java:
```diff
@@ -1264,86 +1274,112 @@ public void testSplitAndTransfer4() {
-      vector.setSafe(7, txt);
+    // Set with convenience function
+    Text txt = new Text("foo");
+    // vector.setSafe(7, txt); cannot use this as a generic method since it is
```
At least, let's not comment out any existing tests.

lidavidm reviewed (commented):
Generally LGTM

vibhatha commented:
@github-actions crossbow submit *java*

vibhatha commented:
@lidavidm I addressed the reviews. And also I added a few additional test cases for views. And earlier there were 4 tests basically covering the view-based logics in storing data. 

Can you think of any uncovered cases here? I am also re-reviewing this. But appreciate your feedback.

vibhatha commented:
`Dev / Source Release and Merge Script on macos-latest ` CI is keep failing. I think there is an issue in installing `Ruby`
cc @kou

kou commented:
#41371
(You can open a new issue for a CI failure. :-)

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java:
```diff
@@ -0,0 +1,1443 @@
+    BitVectorHelper.setValidityBit(validityBuffer, index, isSet);
+    byte[] data = new byte[dataLength];
+    buffer.getBytes(start, data, 0, dataLength);
+    setBytes(index, data, start, dataLength);
```
This still isn't fixed.

vibhatha replied:
Oh I have missed it.

lidavidm commented on java/vector/src/test/java/org/apache/arrow/vector/TestVarCharViewVector.java:
```diff
@@ -0,0 +1,779 @@
+      byte[] view3 = viewVarCharVector.get(2);
+      byte[] view4 = viewVarCharVector.get(3);
+
+      assert view1 != null;
```
use the actual JUnit assert methods, not the assert statement

lidavidm commented on java/vector/src/test/java/org/apache/arrow/vector/TestVarCharViewVector.java:
```diff
@@ -0,0 +1,779 @@
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.util.ReusableByteArray;
+import org.apache.arrow.vector.util.Text;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
```
Let's use JUnit 5 for new tests

lidavidm commented on java/vector/src/test/java/org/apache/arrow/vector/TestVarCharViewVector.java:
Do we have a test of overwriting an existing value? We should test all of overwriting a short string, overwriting a short string with a long string, overwriting a long string with a short string, overwriting a long string with a long string (but shorter) - no reallocation, and overwriting a long string with a longer string - needs a reallocation. (And we should be able to overwrite an element in the middle of the vector in all cases, this is what differentiates this vector from the regular string vector.)

lidavidm replied:
Can we test binary views too?

vibhatha replied:
I will add more tests for binaries, there are some tests.

vibhatha replied:
> Do we have a test of overwriting an existing value? We should test all of overwriting a short string, overwriting a short string with a long string, overwriting a long string with a short string, overwriting a long string with a long string (but shorter) - no reallocation, and overwriting a long string with a longer string - needs a reallocation. (And we should be able to overwrite an element in the middle of the vector in all cases, this is what differentiates this vector from the regular string vector.)

Good point. I will add some tests.

vibhatha replied:
@lidavidm just curious in `ListVector` I observed the following 

```java
try (final VarCharVector vector = new VarCharVector("myviewvector", allocator)) {
      vector.allocateNew(128, 3);

      vector.set(0, "abc".getBytes(StandardCharsets.UTF_8));
      vector.set(1, "012".getBytes(StandardCharsets.UTF_8));
      vector.set(2, "567".getBytes(StandardCharsets.UTF_8));
      vector.set(0, "efghi".getBytes(StandardCharsets.UTF_8));
      vector.setValueCount(3);
      System.out.println(vector);

      Text t1 = vector.getObject(0);
      Text t2 = vector.getObject(1);
      Text t3 = vector.getObject(2);


      System.out.println(t1 + ", " + t1.getLength());
      System.out.println(t2 + ", " + t2.getLength());
      System.out.println(t3 + ", " + t3.getLength());
    }
```

Output

```bash
[efghi, , ]
efghi, 5
, 0
, 0
```

Is this the expected behavior?

vibhatha replied:
Please ignore, this is undefined behavior.

vibhatha replied:
@lidavidm I wrote a few test cases for this. But I have a few doubts, I am going to ask a few questions here, before committing that part. 

### Short with Long

We need to update the view buffer and then append the long string in the data buffer. 

### Long with Less Long

We need to update the view buffer as before, but I have a doubt whether to append this or 
cleanup the existing value and overwrite it? (If I do that there are certain bits which contain unnecessary data, even if I zero it, it will contain unnecessary data unless I remove that bit.) This feels like more compute work, do we need such an approach. Or can we do better? Or do we go with append only approach (which is used at the moment). 

### Long with short 

We need to update the view buffer as before, but do we remove the content in the data buffer which corresponded with the previous insertion?

lidavidm replied:
Long with less long: I think we can overwrite the existing data. We can ignore zeroing of extra bytes for now.

Long with short: we can ignore the content in the data buffer. It's OK to waste a bit of memory for the time being.

lidavidm replied:
And yes with VarCharVector you cannot do any of this. That's why I'm asking for these test cases since this is one of the unique things about this vector.

vibhatha replied:
Will do @lidavidm 
Thanks

vibhatha commented:
@lidavidm I am adding a few more tests.

vibhatha commented:
@github-actions crossbow submit *java*

vibhatha commented:
> #41371 (You can open a new issue for a CI failure. :-)

@kou it seems to be passing now. Was this fixed in another PR?

kou commented:
https://github.com/apache/arrow/pull/41379
